src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Base64Url/Base64UrlDecoder.cs

```diff
@@ -0,0 +1,777 @@
+        }
+
+        /// <summary>
+        /// Decodes the span of unicode ASCII chars represented as Base64Url into binary data.
+        /// </summary>
+        /// <param name="source">The input span which contains unicode ASCII chars in Base64Url that needs to be decoded.</param>
+        /// <param name="destination">The output span which contains the result of the operation, i.e. the decoded binary data.</param>
+        /// <param name="charsConsumed">When this method returns, contains the number of input chars consumed during the operation. This can be used to slice the input for subsequent calls, if necessary. This parameter is treated as uninitialized.</param>
+        /// <param name="bytesWritten">When this method returns, contains the number of bytes written into the output span. This can be used to slice the output for subsequent calls, if necessary. This parameter is treated as uninitialized.</param>
+        /// <param name="isFinalBlock"><see langword="true"/> when the input span contains the entirety of data to encode; <see langword="false"/> when more data may follow,
+        /// such as when calling in a loop, subsequent calls with <see langword="false"/> should end with <see langword="true"/> call. The default is <see langword="true" />.</param>
+        /// <returns>One of the enumeration values that indicates the success or failure of the operation.</returns>
+        /// <remarks>
+        /// As padding is optional the <paramref name="source"/> length not required to be a multiple of 4 even if <paramref name="isFinalBlock"/> is <see langword="true"/>.
+        /// If the <paramref name="source"/> length is not a multiple of 4 and <paramref name="isFinalBlock"/> is <see langword="true"/> the remainders decoded accordingly:
+        /// - Remainder of 3 chars - decoded into 2 bytes data, decoding succeeds.
+        /// - Remainder of 2 chars - decoded into 1 byte data. decoding succeeds.
+        /// - Remainder of 1 char - will cause OperationStatus.InvalidData result.
+        /// </remarks>
+        public static OperationStatus DecodeFromChars(ReadOnlySpan<char> source, Span<byte> destination,
+            out int charsConsumed, out int bytesWritten, bool isFinalBlock = true) =>
+            DecodeFrom<Base64UrlDecoderChar, ushort>(MemoryMarshal.Cast<char, ushort>(source), destination, out charsConsumed, out bytesWritten, isFinalBlock, ignoreWhiteSpace: true);
+
+        private static OperationStatus DecodeWithWhiteSpaceBlockwise<TBase64Decoder>(ReadOnlySpan<ushort> utf8, Span<byte> bytes, ref int bytesConsumed, ref int bytesWritten, bool isFinalBlock = true)
+            where TBase64Decoder : IBase64Decoder<ushort>
+        {
+            const int BlockSize = 4;
+            Span<ushort> buffer = stackalloc ushort[BlockSize];
+            OperationStatus status = OperationStatus.Done;
+
+            while (!utf8.IsEmpty)
```

Comment:
I didn't diff the 3+ copies of this method against each other, but they're all close enough together that it forces me to ask if they can be combined into fewer copies, perhaps with some options to control where they deviate.

Reply:
> I didn't diff the 3+ copies of this method against each other

I believe only have 2 copies of this method, one for `ReadOnlySpan<ushort> utf8`, other for `ReadOnlySpan<byte> utf8`. Others are just wrappers that inside just calling these byte or ushort overload.
 

> but they're all close enough together that it forces me to ask if they can be combined into fewer copies, perhaps with some options to control where they deviate.

Yes, they are identical, I could have used T for byte/short types, but it when set or get span element values I could not cast `T` or `T*` from/to `int` or `byte` any other numeric, in order to do that correctly I need to do type size check for each assignment or access, for example:

```cs
if (sizeof(T) == sizeof(byte))
{
    utf8[index] = (int)Unsafe.Bitcast<T, byte>(src[i]);
}
else
{
    Debug.Assert(sizeof(T) == sizeof(ushort));
    utf8[index] = (int)Unsafe.Bitcast<T, ushort>(src[i]);
}
```
There is quite a lot assignment/access to/from `ReadOnlySpan<T> utf8` in this method, instead I preferred to use the exact type directly with duplicate code, please let me know if you have any other/better idea/preference